receiver/prometheusexecreceiver/receiver_test.go

```diff
@@ -0,0 +1,126 @@
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.uber.org/zap"
+)
+
+// TestEndToEnd loads the test config.yaml and tests two things:
+// 1. makes sure the prometheus_exec config without an exec key throws an error
+// 2. An end-to-end test where metrics are scraped from a fake exporter that exposes Promtheus metrics,
+// and makes sure that exporter subprocess is restarted correctly
+func TestEndToEnd(t *testing.T) {
+	// Load the config from the yaml file
+	factories, err := componenttest.ExampleComponents()
+	assert.NoError(t, err)
+
+	factory := &Factory{}
+	factories.Receivers[factory.Type()] = factory
+
+	config, err := configtest.LoadConfigFile(t, path.Join(".", "testdata", "config.yaml"), factories)
+
+	assert.NoError(t, err)
+	assert.NotNil(t, config)
+
+	// Receiver without exec key, error expected and checked within function
+	execErrorTest(t, config.Receivers["prometheus_exec"])
+
+	// Normal test with port undefined by user
+	endToEndScrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/2"], "end-to-end port not defined")
+}
+
+// execErrorTest makes sure the config passed throws an error, since it's missing the exec key
+func execErrorTest(t *testing.T, errorReceiverConfig configmodels.Receiver) {
+	_, err := new(component.ReceiverCreateParams{Logger: zap.NewNop()}, errorReceiverConfig.(*Config), nil)
+	if err == nil {
+		t.Errorf("end_to_end_test.go didn't get error, was expecting one since this config has no 'exec' key")
+	}
+}
+
+// endToEndScrapeTest scrapes a test endpoint (test_prometheus_exporter.go) twice, and between each scrape yields the execution with Sleep() to wait for the subprocess (exporter) to restar
+// - wait time is about 1s - to fail and restart, meaning it verifies three things: the scrape is successful (twice), the process was restarted correctly when failed and the underlying
+// Prometheus receiver was correctly stopped and then restarted. For extra testing the metrics values are different every time the subprocess exporter is started
+// And the uniqueness of the metric scraped is verified
+func endToEndScrapeTest(t *testing.T, receiverConfig configmodels.Receiver, testName string) {
+	sink := &exportertest.SinkMetricsExporter{}
+	wrapper, err := new(component.ReceiverCreateParams{Logger: zap.NewNop()}, receiverConfig.(*Config), sink)
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %w", err)
+	}
+
+	ctx := context.Background()
+	err = wrapper.Start(ctx, componenttest.NewNopHost())
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %w", err)
+	}
+	defer wrapper.Shutdown(ctx)
+
+	var metrics []pdata.Metrics
+
+	// Make sure a first scrape works by checking for metrics in the test metrics exporter "sink", only return true when there are metrics
+	const waitFor = 15 * time.Second
+	const tick = 500 * time.Millisecond
```

Comment:
Lowering tick, but leaving high `waitFor` since test seems to take about 6-8s on fast machine, anyways the `eventually` will be exited when it returns true.